Add a naked pointers dynamic checker, continued

xavierleroy opened the pull request:
This is a variant of #9947 where the check for "naked pointers" (dangerous out-of-heap pointers) is performed in a runtime system that supports naked pointers and will not crash early when it encounters one.

Also, the check is disabled in bytecode, as we don't catch SEGV signals in this case.

Tests were added to the test suite that exercise the three modes: naked pointers / naked pointers + dynamic checker / no naked pointers.

kayceesrk committed:
Initial commit of naked pointer checker

kayceesrk committed:
Atoms have black header. Objects must be reasonable sized.

kayceesrk committed:
Disable optimisations for is_pointer_safe

kayceesrk committed:
Convince clang to generate the correct code

kayceesrk committed:
Use inline assembly to circumvent obstinate clang

dra27 committed:
Stephen's safe_load code for msvc64

dra27 committed:
SIZEOF_PTR != SIZEOF_LONG on Windows

abbysmal committed:
nnp checker: cherry-pick to trunk, update to support 4.12.0 and configure additions

abbysmal committed:
Changes entry for naked pointers checker

abbysmal committed:
nnp checker: checking_pointer_pc not used on _WIN32

abbysmal committed:
update Changes entry

xavierleroy committed:
Revise the checker for naked pointers

- It must run in naked pointers mode, otherwise other parts of the runtime
  system will crash on naked pointers before they can be detected
- It cannot run in bytecode because catching SIGSEGV is done only in
  native code

xavierleroy committed:
--enable-naked-pointers-checker is incompatible with --disable-naked-pointers

xavierleroy committed:
Tests for naked pointer support and naked pointer checking

np1 and np2 are valid uses of out-of-heap pointers even in
no-naked-pointers mode.  np2 will generate an alarm in checking mode, though.

np3 and np4 are OK only in naked-pointers mode.  They will generate alarms
in checking mode.

xavierleroy committed:
Kneel to check-typo

kayceesrk commented on runtime/major_gc.c:
```diff
@@ -1077,3 +1088,89 @@ void caml_finalise_heap (void)
+  /* For the pointer to be considered safe, either the given pointer is in heap
+   * or the (out of heap) pointer has a black header and its size is < 2 ** 40
+   * words (128 GB). If not, we report a warning. */
+  if (Is_in_heap (v) ||
```
IIUC `check_pointer_safe` is only invoked when `!Is_in_heap(v) && !Is_young(v)`. `check_pointer_safe` should not be called on a pointer in the minor heap since OCaml does not define what the colour of minor objects is.

Would it be useful to rename this function to `caml_is_naked_pointer_safe`, with an assertion `CAMLassert (!Is_in_heap(v) && !Is_young(v))` to make these assumptions explicit?

xavierleroy replied:
Yes, you're absolutely right!  I renamed to `is_naked_pointer_safe` (it's a static function, so no need for `caml_`), added the assertion, and removed the `Is_in_heap` test within.

kayceesrk commented:
I've reviewed the changes and it looks good to me. In particular, having the test in naked pointers mode makes sense to me.

nojb commented:
@kayceesrk Just to make sure I understand, in [this comment](https://github.com/ocaml/ocaml/pull/9947#issuecomment-703045740) it is written that `safe_load` is no longer needed, but the function is still present in the code. Did I misunderstand the comment?

kayceesrk commented:
I should have clarified that `safe_load` was no longer needed _in `mark_stack_push`_ since we no longer load the header since #9951. See https://github.com/ocaml/ocaml/pull/9951/files#diff-554731eb45f6f98a2ad012dfbe66f9caL244.

nojb commented:
Thanks for the explanation, I understand now.

xavierleroy committed:
Take review comments into account

- Rename `check_pointer_safe` -> `is_naked_pointer_safe`
- Make it clear that it is called only on out-of-heap pointers
  (neither Is_young nor Is_in_heap), as checked by the caller.

kayceesrk commented:
Lgtm!
